algorithm/cgal/Fair.h

```diff
@@ -0,0 +1,130 @@
+    selections.push_back(&geometry->epick_mesh(selection));
+  }
+
+  for (size_t nth = 0; nth < count; nth++) {
+    if (!geometry->is_mesh(nth)) {
+      continue;
+    }
+    Epick_surface_mesh& mesh = geometry->epick_mesh(nth);
+
+    std::vector<Epick_surface_mesh::Face_index> selected_faces;
+
+    if (count < size) {
+      std::vector<const Epick_surface_mesh*> selections;
+
+      for (int selection = count; selection < size; selection++) {
+        // Add seams at the selection boundaries.
+        std::cout << "Corefining" << std::endl;
+        CGAL::Polygon_mesh_processing::corefine(
+            mesh, geometry->epick_mesh(selection),
+            CGAL::parameters::all_default(), CGAL::parameters::all_default());
+        selections.push_back(&geometry->epick_mesh(selection));
+      }
+    }
+
+    if (count < size) {
+      std::cout << "Selecting faces" << std::endl;
+      // Select all faces not extending outside of the selection.
+      // Since the mesh was seamed these will touch the selection boundary.
+      for (Epick_surface_mesh::Face_index face : mesh.faces()) {
+        Halfedge_index start = mesh.halfedge(face);
+        Halfedge_index edge = start;
+        bool outside = false;
+        do {
+          const Epick_point& p = mesh.point(mesh.source(edge));
+          for (size_t selection = count; selection < size; selection++) {
+            if (geometry->on_epick_side(selection)(p) ==
+                CGAL::ON_UNBOUNDED_SIDE) {
+              // This face must not be refined.
+              outside = true;
+              break;
+            }
+          }
+          edge = mesh.next(edge);
+        } while (edge != start && !outside);
+        if (!outside) {
+          selected_faces.push_back(face);
+        }
+      }
+    } else {
+      for (Epick_surface_mesh::Face_index face : mesh.faces()) {
+        selected_faces.push_back(face);
+      }
+    }
+
+    std::cout << "Selected face count: " << selected_faces.size() << std::endl;
+
+    {
+      std::vector<Epick_surface_mesh::Face_index> new_facets;
+      std::vector<Epick_surface_mesh::Vertex_index> new_vertices;
+      MakeDeterministic();
+      CGAL::Polygon_mesh_processing::refine(
+          mesh, selected_faces, std::back_inserter(new_facets),
+          std::back_inserter(new_vertices),
+          CGAL::parameters::density_control_factor(density));
+      std::cout << "New face count: " << new_facets.size() << std::endl;
+      std::cout << "New vertex count: " << new_vertices.size() << std::endl;
+    }
+
+    std::vector<Epick_surface_mesh::Vertex_index> selected_vertices;
+
+    if (count < size) {
+      for (Epick_surface_mesh::Vertex_index vertex : mesh.vertices()) {
+        for (size_t selection = count; selection < size; selection++) {
+          if (geometry->on_epick_side(selection)(mesh.point(vertex)) ==
+              CGAL::ON_BOUNDED_SIDE) {
+            // Only vertices within the section can be moved.
```

Comment:
How does selection work?

Reply:
You provide shapes which delineate the volume to be faired.

e.g., Box(10, 10, 10).fair(Box(11, 11, [6])) should fair the top half of the box.

Unfortunately fairing doesn't really work as nicely as I was hoping -- I can't figure out if it's because my expectations are wrong, or if something's off in the execution.